NetworkCORSPreflightChecker: Ask for smartcard PIN.

mitchdz opened the pull request:
#### 3fd7303a942afa28c5a65df53ae892d9910afbf5
<pre>
NetworkCORSPreflightChecker: Ask for smartcard PIN.
<a href="https://bugs.webkit.org/show_bug.cgi?id=271183.">https://bugs.webkit.org/show_bug.cgi?id=271183.</a>

Reviewed by Michael Catanzaro.

If a smartcard is PIN protected, attempting to use it with webkit2gtk
will not work as the PIN is not requested. This patch fixes that by
asking for a PIN.

* Source/WebKit/NetworkProcess/NetworkCORSPreflightChecker.cpp:
(WebKit::NetworkCORSPreflightChecker::didReceiveChallenge): ask for
credential PIN

Canonical link: <a href="https://commits.webkit.org/278811@main">https://commits.webkit.org/278811@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7681d3fd92f52f6dc7621d55529b0a3ef92ef5df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51629 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54895 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2321 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37296 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2005 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42027 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53728 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23155 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25888 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47851 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1885 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56487 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49427 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27987 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44604 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/48647 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11302 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28884 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27724 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (changes requested):
Need to fix the build failure for non-glib case

mcatanzaro commented on Source/WebKit/NetworkProcess/NetworkCORSPreflightChecker.cpp:
```diff
@@ -103,7 +103,13 @@ void NetworkCORSPreflightChecker::didReceiveChallenge(WebCore::AuthenticationCha
+    if (!askUserForCredentials && !isTLSHandshake) {
+#else
     if (!isTLSHandshake) {
+#endif
```
I'm going to suggest writing this differently to avoid mixing `#if` with `if`:

```suggestion
#if USE(GLIB)
    bool askUserForCredentials = scheme == ProtectionSpace::AuthenticationScheme::ClientCertificatePINRequested;
#else
    bool askUserForCredentials = false;
#endif

    if (!askUserForCredentials && !isTLSHandshake) {
```

mitchdz replied:
Great suggestion, I like this a ton better too. Will add that now.

mitchdz committed:
NetworkCORSPreflightChecker: Ask for smartcard PIN.
https://bugs.webkit.org/show_bug.cgi?id=271183.

Reviewed by Michael Catanzaro.

If a smartcard is PIN protected, attempting to use it with webkit2gtk
will not work as the PIN is not requested. This patch fixes that by
asking for a PIN.

* Source/WebKit/NetworkProcess/NetworkCORSPreflightChecker.cpp:
(WebKit::NetworkCORSPreflightChecker::didReceiveChallenge): ask for
credential PIN

Canonical link: https://commits.webkit.org/278811@main
